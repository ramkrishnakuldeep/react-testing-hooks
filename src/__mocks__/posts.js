export default [{
    "createdAt": "2019-08-02T06:43:34.050Z",
    "id": "e9cabd15-602b-43eb-85be-b17b82c74983",
    "userId": 1,
    "body": "1234567890",
    "title": "09090909"
}, {
    "createdAt": "2019-08-01T23:40:39.885Z",
    "id": "4f6fb244-931b-453f-9a19-9c779b610c68",
    "userId": 1,
    "title": "make this go away",
    "body": "the placeholder remaining is annoying"
}, {
    "createdAt": "2019-08-01T03:28:40.867Z",
    "id": "ea31bef4-5144-41d1-8832-344815fec140",
    "userId": 1,
    "body": " what happens",
    "title": "test with materialize"
}, {
    "createdAt": "2019-08-01T03:05:49.730Z",
    "id": "ef6b63e5-f003-4a2c-97da-ec2338cdba98",
    "userId": 1,
    "body": "buckley",
    "title": "alanb"
}, {
    "createdAt": "2019-07-31T18ÃŽ:01:17.926Z",
    "id": "c13daef7-2ec6-433f-a050-964377c24a06",
    "userId": 1,
    "body": "sas",
    "title": "sis"
}, {
    "createdAt": "2019-07-31T18:00:05.635Z",
    "id": "13fb0d20-d461-4b73-9809-793d62031939",
    "userId": 1,
    "title": "New post from Vue",
    "body": "Vue is cool"
}, {
    "createdAt": "2019-07-31T14:41:23.889Z",
    "id": "85ba50b4-005b-49ab-a5ec-ad5d431b3318",
    "userId": 1,
    "title": "new post from the app",
    "body": "Hello new post!"
}, {
    "createdAt": "2019-07-30T21:51:28.835Z",
    "id": "d9c21d8f-0feb-4e04-be24-bf3cd8d4bf72",
    "userId": 1,
    "body": "May you always find solace from the light, and peace inside your mind",
    "title": "Darkness"
}]

