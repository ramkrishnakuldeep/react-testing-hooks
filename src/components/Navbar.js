import React from 'react';

import { Link } from 'react-router-dom';

function Navbar() {
    return (
        <nav>
            <div className="nav-wrapper">
                <Link to="/" className="brand-logo">Hooks Demo</Link>
                <ul id="nav-mobile" className="right hide-on-med-and-down">
                    <li><Link data-testid='nav-home' to="/">Home</Link></li>
                    <li><Link data-testid='nav-about' to="/about">About</Link></li>
                </ul>
            </div>
        </nav>
    )
}

export default Navbar
