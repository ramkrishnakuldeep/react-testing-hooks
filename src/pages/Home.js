import React, {useEffect, useState} from 'react'
import axios from 'axios';
import PostForm from '../components/PostForm';

const Home = () => {
    const [posts, setPosts] = useState([]);
    const [limit, setLimit] = useState(5);
    const [editingPost, setEditingPost] = useState({title: '', body: '', id: null});

    async function fetchPosts() {
        let response = await axios.get('/posts');
        console.log(response.data);
        setPosts(response.data);
    }
    
    useEffect(() => {
        fetchPosts();
    }, []);

    function editPost(post) {
        console.log(post);
        setEditingPost(post);

    }

    function deletePost(id) {
        axios.delete(`/post/${id}`).then(() => {
            const updatedPosts = posts.filter(p => p.id !== id);
            setPosts(updatedPosts);
        }).catch(error => {
            console.error(error);
        })
    }

    const getNumberOfPosts = () => {
        axios.get(`/posts/${limit}`).then(res => {
            setPosts(res.data)
        }).catch(err => {
            console.error(err);
        })
    }
    const addPost = (post) => {

        if(posts.find(p => p.id === post.id)){
            const index = posts.findIndex(p => p.id === post.id);
            const updatedPosts = [...posts];
            updatedPosts.splice(index, 1, post);
            setPosts(updatedPosts);
        } else {
            const updatedPosts = [...posts, post];
            setPosts(updatedPosts);
        }

        
    }

    return (
        <div>
            <div className='row'>
                <div className='col s6'>
                    <PostForm addPost={addPost} editingPost={editingPost} />
                </div>
                <div className="col s3 push-in">
                    <p>
                        Limit number of posts
                    </p>
                    <input type="number" value={limit} name="limit" onChange={event => setLimit(event.target.value)} />
                    <button className="waves-effect waves-light btn" onClick={getNumberOfPosts}>
                        Set
                    </button>
                </div>
            </div>
            
            <div className='row'>
                {
                    posts.map(post => (
                        <div className='col s6' key={post.id}>
                            <div className='card'>
                                <div className='card-content'>

                                
                                <div className='card-title'>
                                    {post.title}
                                </div>
                                <p className='timestamp'>
                                    {post.createdAt}
                                </p>
                                <p className='card-body'>
                                    {post.body}
                                </p>
                                
                                </div>
                                <div className='card-action'> 
                                    <a onClick={editPost.bind(null, post)}>
                                        Edit
                                    </a>
                                    <a className='delete-btn' onClick={deletePost.bind(null, post.id)}>
                                        Delete
                                    </a>
                                    
                                </div>
                            </div>
                        </div>
                    ))
                }
                
            </div>
        </div>
    )
}

export default Home
