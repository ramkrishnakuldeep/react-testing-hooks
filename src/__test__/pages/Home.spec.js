import React from 'react';
import { render, cleanup, act, waitForElement } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import axiosMock from 'axios';
import Home from '../../pages/Home';
import Navbar from '../../components/Navbar'
import posts from '../../__mocks__/posts';


describe('Home component ', () => {
    afterEach(cleanup);

    it('render without crash', async() => {
        axiosMock.get.mockResolvedValueOnce({
            data: posts
        });
        let wrapper;
        act(() => {
            wrapper = render(<Home />);
        });
        await waitForElement(() => wrapper.asFragment());
        expect(wrapper.asFragment()).toMatchSnapshot();
    })
})
