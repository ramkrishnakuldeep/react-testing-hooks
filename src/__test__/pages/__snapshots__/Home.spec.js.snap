// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component  render without crash 1`] = `
<DocumentFragment>
  <div>
    <div
      class="row"
    >
      <div
        class="col s6"
      >
        <form
          class="push-in"
        >
          <div
            class="input-field"
          >
            <label
              for="title"
            >
              Title
            </label>
            <input
              name="title"
              type="text"
              value=""
            />
            <span
              class="helper-text"
            />
          </div>
          <div
            class="input-field"
          >
            <label
              for="body"
            >
              Body
            </label>
            <input
              name="body"
              type="text"
              value=""
            />
            <span
              class="helper-text"
            />
          </div>
          <button
            class="waves-effect waves-light btn"
            type="submit"
          >
            Add
          </button>
        </form>
      </div>
      <div
        class="col s3 push-in"
      >
        <p>
          Limit number of posts
        </p>
        <input
          name="limit"
          type="number"
          value="5"
        />
        <button
          class="waves-effect waves-light btn"
        >
          Set
        </button>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col s6"
      >
        <div
          class="card"
        >
          <div
            class="card-content"
          >
            <div
              class="card-title"
            >
              09090909
            </div>
            <p
              class="timestamp"
            >
              2019-08-02T06:43:34.050Z
            </p>
            <p
              class="card-body"
            >
              1234567890
            </p>
          </div>
          <div
            class="card-action"
          >
            <a>
              Edit
            </a>
            <a
              class="delete-btn"
            >
              Delete
            </a>
          </div>
        </div>
      </div>
      <div
        class="col s6"
      >
        <div
          class="card"
        >
          <div
            class="card-content"
          >
            <div
              class="card-title"
            >
              make this go away
            </div>
            <p
              class="timestamp"
            >
              2019-08-01T23:40:39.885Z
            </p>
            <p
              class="card-body"
            >
              the placeholder remaining is annoying
            </p>
          </div>
          <div
            class="card-action"
          >
            <a>
              Edit
            </a>
            <a
              class="delete-btn"
            >
              Delete
            </a>
          </div>
        </div>
      </div>
      <div
        class="col s6"
      >
        <div
          class="card"
        >
          <div
            class="card-content"
          >
            <div
              class="card-title"
            >
              test with materialize
            </div>
            <p
              class="timestamp"
            >
              2019-08-01T03:28:40.867Z
            </p>
            <p
              class="card-body"
            >
               what happens
            </p>
          </div>
          <div
            class="card-action"
          >
            <a>
              Edit
            </a>
            <a
              class="delete-btn"
            >
              Delete
            </a>
          </div>
        </div>
      </div>
      <div
        class="col s6"
      >
        <div
          class="card"
        >
          <div
            class="card-content"
          >
            <div
              class="card-title"
            >
              alanb
            </div>
            <p
              class="timestamp"
            >
              2019-08-01T03:05:49.730Z
            </p>
            <p
              class="card-body"
            >
              buckley
            </p>
          </div>
          <div
            class="card-action"
          >
            <a>
              Edit
            </a>
            <a
              class="delete-btn"
            >
              Delete
            </a>
          </div>
        </div>
      </div>
      <div
        class="col s6"
      >
        <div
          class="card"
        >
          <div
            class="card-content"
          >
            <div
              class="card-title"
            >
              sis
            </div>
            <p
              class="timestamp"
            >
              2019-07-31T18ÃŽ:01:17.926Z
            </p>
            <p
              class="card-body"
            >
              sas
            </p>
          </div>
          <div
            class="card-action"
          >
            <a>
              Edit
            </a>
            <a
              class="delete-btn"
            >
              Delete
            </a>
          </div>
        </div>
      </div>
      <div
        class="col s6"
      >
        <div
          class="card"
        >
          <div
            class="card-content"
          >
            <div
              class="card-title"
            >
              New post from Vue
            </div>
            <p
              class="timestamp"
            >
              2019-07-31T18:00:05.635Z
            </p>
            <p
              class="card-body"
            >
              Vue is cool
            </p>
          </div>
          <div
            class="card-action"
          >
            <a>
              Edit
            </a>
            <a
              class="delete-btn"
            >
              Delete
            </a>
          </div>
        </div>
      </div>
      <div
        class="col s6"
      >
        <div
          class="card"
        >
          <div
            class="card-content"
          >
            <div
              class="card-title"
            >
              new post from the app
            </div>
            <p
              class="timestamp"
            >
              2019-07-31T14:41:23.889Z
            </p>
            <p
              class="card-body"
            >
              Hello new post!
            </p>
          </div>
          <div
            class="card-action"
          >
            <a>
              Edit
            </a>
            <a
              class="delete-btn"
            >
              Delete
            </a>
          </div>
        </div>
      </div>
      <div
        class="col s6"
      >
        <div
          class="card"
        >
          <div
            class="card-content"
          >
            <div
              class="card-title"
            >
              Darkness
            </div>
            <p
              class="timestamp"
            >
              2019-07-30T21:51:28.835Z
            </p>
            <p
              class="card-body"
            >
              May you always find solace from the light, and peace inside your mind
            </p>
          </div>
          <div
            class="card-action"
          >
            <a>
              Edit
            </a>
            <a
              class="delete-btn"
            >
              Delete
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
